core/modules/wsched/ChunkTaskList.cc

```diff
@@ -0,0 +1,393 @@
+    if (_activeChunk == _chunkMap.end()) {
+        return true;
+    }
+    return _activeChunk->second->readyToAdvance();
+}
+
+
+void ChunkTaskList::taskComplete(wbase::Task::Ptr const& task) {
+    auto iter = _chunkMap.find(task->getChunkId());
+    if (iter != _chunkMap.end()) {
+        iter->second->taskComplete(task);
+    }
+}
+
+
+bool ChunkTaskList::setResourceStarved(bool starved) {
+    bool ret = _resourceStarved;
+    _resourceStarved = starved;
+    return ret;
+}
+
+
+void ChunkTasks::SlowTableHeap::push(wbase::Task::Ptr const& task) {
+    _tasks.push_back(task);
+    std::push_heap(_tasks.begin(), _tasks.end(), compareFunc);
+}
+
+
+wbase::Task::Ptr ChunkTasks::SlowTableHeap::pop() {
+    if (_tasks.empty()) {
+        return nullptr;
+    }
+    auto task = _tasks.front();
+    std::pop_heap(_tasks.begin(), _tasks.end(), compareFunc);
+    _tasks.pop_back();
+    return task;
+}
+
+
+/// Queue new Tasks to be run, ordered with the slowest tables first.
+/// This relies on ChunkTasks owner for thread safety.
+void ChunkTasks::queTask(wbase::Task::Ptr const& a) {
+    time(&a->entryTime);
+    /// Compute entry time to reduce spurious valgrind errors
+    ::ctime_r(&a->entryTime, a->timestr);
+
+    const char* state = "";
+    // If this is the active chunk, put new Tasks on the pending list, as
+    // we could easily get stuck on this chunk as new Tasks come in.
+    if (_active) {
+        _pendingTasks.push_back(a);
+        state = "PENDING";
+    } else {
+        _activeTasks.push(a);
+        state = "ACTIVE";
+    }
+    LOGS(_log, LOG_LVL_DEBUG,
+         "ChunkTasks enqueue "
+         << a->getIdStr()
+         << " chunkId=" << _chunkId
+         << " state=" << state
+         << " active.sz=" << _activeTasks._tasks.size()
+         << " pend.sz=" << _pendingTasks.size());
+    if (_activeTasks.empty()) {
+        LOGS(_log, LOG_LVL_DEBUG, "Top of ACTIVE is now: (empty)");
+    } else {
+        LOGS(_log, LOG_LVL_DEBUG, "Top of ACTIVE is now: " << _activeTasks.top()->getIdStr());
+    }
+}
+
+
+/// Set this chunk as the active chunk and move pending jobs to active if needed.
+void ChunkTasks::setActive(bool active) {
+    if (_active != active) {
+        LOGS(_log, LOG_LVL_DEBUG, "ChunkTasks " << _chunkId << " active changed to " << active);
+        if (_active && !active) {
+            movePendingToActive();
+        }
+    }
+    _active = active;
+}
+
+
+/// Move all pending Tasks to the active heap.
+void ChunkTasks::movePendingToActive() {
+    for (auto const& t:_pendingTasks) {
+        LOGS(_log, LOG_LVL_DEBUG, "ChunkTasks " << _chunkId << " pending->active " << t->getIdStr());
+        _activeTasks.push(t);
+    }
+    _pendingTasks.clear();
+}
+
+
+/// @return true if active AND pending are empty.
+bool ChunkTasks::empty() const {
+    return _activeTasks.empty() && _pendingTasks.empty();
+}
+
+
+/// This is ready to advance when _activeTasks is empty and no Tasks are in flight.
```

Comment:
Does this mean that once a task becomes active you can't add any more? What avoids always running a single task just because you advanced something too quickly?